### ISSUE TITLE
`AsyncAccept` is not implemented for `AddrIncoming` || `accept()` cannot be called on `TlsListener<AddrIncoming, TlsAcceptor>` due to unsatisfied trait bounds

### ISSUE THREAD

**Issue by robertpauson** [tmccombs/tls-listener#30]
From examples:
```

fn tls_acceptor_impl(cert_der: &[u8], key_der: &[u8]) -> tokio_rustls::TlsAcceptor {
    let key = PrivateKey(cert_der.into());
    let cert = Certificate(key_der.into());
    Arc::new(
        ServerConfig::builder()
            .with_safe_defaults()
            .with_no_client_auth()
            .with_single_cert(vec![cert], key)
            .unwrap(),
    )
    .into()
}

fn tls_acceptor() -> tokio_rustls::TlsAcceptor {
    tls_acceptor_impl(PKEY, CERT)
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let addr = SocketAddr::from(...);
    let mut listener = tls_listener::builder(tls_acceptor())
        .max_handshakes(10)
        .listen(AddrIncoming::bind(&addr).unwrap());

    let (tx, mut rx) = mpsc::channel::<tokio_rustls::TlsAcceptor>(1);

    let http = Http::new();
    loop {
        tokio::select! {
            conn = listener.accept() => {
                match conn.expect("Tls listener stream should be infinite") {
                    Ok(conn) => {
                        let http = http.clone();
                        // let tx = tx.clone();
                        // let counter = counter.clone();
                        tokio::spawn(async move {
                            // let svc = service_fn(move |request| handle_request(tx.clone(), counter.clone(), request));
                            if let Err(err) = http.serve_connection(conn, service_fn(my_query_handler)).await {
                                eprintln!("Application error: {}", err);
                            }
                        });
                    },
                    Err(e) => {
                        eprintln!("Bad connection: {}", e);
                    }
                }
            },
            message = rx.recv() => {
                // Certificate is loaded on another task; we don't want to block the listener loop
                let acceptor = message.expect("Channel should not be closed");
            }
        }
    }
```
Produces error in title. 

A different approach using this instead of `loop{}` from another example:

```
    listener
        .for_each(|r| async {
            match r {
                Ok(conn) => {
                    let http = http.clone();
                    tokio::spawn(async move {
                        if let Err(err) = http.serve_connection(conn, service_fn(my_query_handler)).await {
                            eprintln!("Application error: {}", err);
                        }
                    });
                }
                Err(err) => {
                    eprintln!("Error accepting connection: {}", err);
                }
            }
        })
        .await;
```
Exact same error. On `listen()` and `.for_each()`.

My Cargo.toml:
```
[dependencies]
bson = "2.4.0"
futures = "0.3.25"
futures-util = "0.3.25"
hyper = {version="0.14.23",  features = ["http1", "server", "tcp"]}
mongodb = "2.3.1"
rand = "0.8.5"
rustls = "0.20.7"
serde = "1.0.150"
serde_json = "1.0.89"
tls-listener = {version="0.5.1", features=["rustls"]}
tokio = "1.23.0"
tokio-rustls = "0.23.4"
url = "2.3.1"

```

From chatgpt:

```
It looks like you are trying to create a TLS listener with tls_listener::builder. This function returns a TlsListener<AddrIncoming, TlsAcceptor>, which is a type that implements the AsyncRead and AsyncWrite traits, but not the AsyncAccept trait.

The error message "AsyncAccept is not implemented for AddrIncoming" suggests that the listener object you created is not a valid acceptor, which means that you cannot call the accept() method on it.

One way to fix this error would be to use a different type of listener that does implement the AsyncAccept trait. For example, you could use the tokio::net::TcpListener type to create a TCP listener and then wrap it in a TlsListener using the tokio_rustls::TlsAcceptor::new method.
```

its answer is wrong and results in same error
```
    let addr = SocketAddr::from(...);
    let tcp_listener = TcpListener::bind(&addr).expect("Failed to bind TCP listener");
    let tls_acceptor = tls_acceptor();
    let mut listener = TlsListener::new(tcp_listener, tls_acceptor);

    loop {
        let (conn, _) = listener.accept().await?;
        let http = http.clone();
        tokio::spawn(async move {
            if let Err(err) = http.serve_connection(conn, service_fn(my_query_handler)).await {
                eprintln!("Application error: {}", err);
            }
        });
    }
```

it ommited `Http::new()` and upon fixing that the errors are abundant:

```
error[E0277]: the trait bound `TlsAcceptor: AsyncAccept` is not satisfied
   --> src/main.rs:374:55
    |
374 |     let mut listener = TlsListener::new(tcp_listener, tls_acceptor);
    |                        ----------------               ^^^^^^^^^^^^ the trait `AsyncAccept` is not implemented for `TlsAcceptor`
    |                        |
    |                        required by a bound introduced by this call
```
```
error[E0277]: the trait bound `TlsAcceptor: AsyncAccept` is not satisfied
   --> src/main.rs:374:24
    |
374 |     let mut listener = TlsListener::new(tcp_listener, tls_acceptor);
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `AsyncAccept` is not implemented for `TlsAcceptor`
    |
```

```
error[E0277]: the trait bound `tokio::net::TcpListener: AsyncTls<_>` is not satisfied
   --> src/main.rs:374:24
    |
374 |     let mut listener = TlsListener::new(tcp_listener, tls_acceptor);
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `AsyncTls<_>` is not implemented for `tokio::net::TcpListener`
    |
    = help: the trait `AsyncTls<C>` is implemented for `TlsAcceptor`
```

```
error[E0599]: the method `accept` exists for struct `TlsListener<TlsAcceptor, tokio::net::TcpListener>`, but its trait bounds were not satisfied
   --> src/main.rs:377:34
    |
377 |         let (conn, _) = listener.accept().await?;
    |                                  ^^^^^^ method cannot be called on `TlsListener<TlsAcceptor, tokio::net::TcpListener>` due to unsatisfied trait bounds
    |
```

**Comment by tmccombs** [comment]
You need to enable at least one of the hyper-h1 or hyper-h2 features in order to get the implementation for AddrIncoming

**Comment by robertpauson** [comment]
> You need to enable at least one of the hyper-h1 or hyper-h2 features in order to get the implementation for AddrIncoming
```the package `api-test` depends on `tls-listener`, with features: `rustls, hyper-h2, hyper-h1` but `tls-listener` does not have these features.```
```the package `alpha-rest-api` depends on `tls-listener`, with features: `rustls, hyper-h1` but `tls-listener` does not have these features.```

```the package `alpha-rest-api` depends on `tls-listener`, with features: `rustls, hyper-h2` but `tls-listener` does not have these features.```

only `hyper-h1`:
```389 |         .listen(AddrIncoming::bind(&addr).unwrap());
    |          ^^^^^^ the trait `AsyncTls<_>` is not implemented for `TlsAcceptor`
```

only `hyper-h2`:
```389 |         .listen(AddrIncoming::bind(&addr).unwrap());
    |          ^^^^^^ the trait `AsyncTls<_>` is not implemented for `TlsAcceptor`
    |```

**Comment by tmccombs** [comment]
I'm not sure what is causing the first error you listed. I just tried making a test project with `tls-listener = {version="0.5.1", features=["rustls", "hyper-h1"]}` , and it built fine.

**Comment by robertpauson** [comment]
Somehow, suddenly the error stopped appearing, and it compiled! Closing.

Apologies in advance if that's not the right place for a follow-up question, but now I am a bit confused about my SSL Certs.
Calling my API now ported to HTTPS, with python requests I'm getting `SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)'`. 

The keys I used are my `certbot` generated `pem` keys converted to `der` format using those commands:

`openssl x509 -outform der -in /etc/letsencrypt/live/mysite.com/cert.pem -out /etc/letsencrypt/live/mysite.com/cert.der`

`x509` on `privkey` was giving me errors but the file is not corrupt... so I went for:

`openssl rsa -in /etc/letsencrypt/live/mysite.com/privkey.pem -outform der -out /etc/letsencrypt/live/mysite.com/privkey.der`

But I am not sure whether that can even work as docs for tokio_rustls say it needs to be `x509`. Well it compiles, polls... and just throws this error from example - IDK what this might mean, I have 0 experience with this. 
`Bad connection: cannot decrypt peer's message`
